plutus-core/index-envs/src/Data/RandomAccessList/RelativizedMap.hs

```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Data.RandomAccessList.RelativizedMap (RelativizedMap (..))where
+
+import Data.Word
+
+import Data.RandomAccessList.Class qualified as RAL
+
+import Data.IntMap.Strict qualified as IM
+import GHC.Exts (IsList)
+
+-- | A sequence implemented by a map from "levels" to values and a counter giving the "current" level.
+data RelativizedMap a = RelativizedMap (IM.IntMap a) {-# UNPACK #-} !Word64
+    deriving stock (Show, Eq)
+    deriving (IsList) via RAL.AsRAL (RelativizedMap a)
+
+instance RAL.RandomAccessList (RelativizedMap a) where
+    type Element (RelativizedMap a) = a
+
+    {-# INLINABLE empty #-}
+    empty = RelativizedMap mempty 0
+    {-# INLINABLE cons #-}
+    cons a (RelativizedMap im l) = RelativizedMap (IM.insert (fromIntegral l) a im) (l+1)
+    {-# INLINABLE uncons #-}
+    uncons (RelativizedMap _ 0) = Nothing
```

Comment:
now that i am thinking about it, this case is redundant because we guarantee that when the cached length / level is 0 then the map is empty

Reply:
That's true, but I think this way is clearer :)